doc/console/Portable-Console-Device-High-Level-Design.md

```diff
@@ -0,0 +1,390 @@
+        except Exception as e:
+            return (None, e)
+
+    def get_serial_number(self):
+        return "Microsoft-Simulator-S/N"
+
+    def get_virtual_device_list(self):
+        return {
+            1: "/dev/console-1",
+            2: "/dev/console-2",
+            # ...
+        }
+
+    def get_num_psus(self):
+        return 0
+
+    def get_all_psus(self):
+        return []
+
+    def get_psu(self, index):
+        return None
+```
+
+### Factory Function Design
+
+Factory function `get_portable_console_device` supports three ways to identify which subclass should be used to create portable console device object:
+
+1. Use function parameter `vendor_name` and `model_name` to specify which subclass to use. (This can be used to write unit test.)
+2. If `autodetect` is set to `disable` in `config_db`, then use `vendor_name` in `config_db` to specify which subclass to use. (If a vendor doesn't support the `is_plugged_in` method, the device can be manually confitured to work.)
+3. If `autodetect` is set to `enable`, then use `is_plugged_in` method of all the derived subclass to detect which vendor's device is plugged in, and use the corresponding subclass. (**recommended**)
+
+The flow chart below describes how `get_portable_console_device` function works:
+
+![factory-function-flow-chart.png](./Portable-Console-Device-High-Level-Design/factory-function.png)
+
+As mentioned above, only the third way is our recommendation, which can automatically detect which vendor's device is plugged in and create the corresponing object. The first and second ways are reserved for more flexibility, so they are given higher priority.
+
+## SONiC CLI Design
+
+### Console Show Commands
+
+#### show console summary
+
+This command displays a summary of the portable console device.
+
+* Usage:
+
+  ```
+  show console summary
+  ```
+
+* Example:
+
+  ```
+  admin@sonic:~$ show console summary
+  Vendor Name: Microsoft
+  Model Name: Simulator
+  Serial Number: Microsoft-Simulator-S/N
+  Auto Detect: Disable
+  ```
+
+#### show console vendor_name
+
+This command displays the vendor name of the portable console device. If auto-detect is disabled, it will read the vendor name from CONFIG_DB. Otherwise, it will read the vendor name from `get_vendor_name()`.
+
+* Usage:
+
+  ```
+  show console vendor_name
+  ```
+
+* Example:
+
+  ```
+  admin@sonic:~$ show console vendor_name
+  Microsoft
+  ```
+
+#### show console model_name
+
+This command displays the model name of the portable console device. If auto-detect is disabled, it will read the model name from CONFIG_DB. Otherwise, it will read the model name from `get_vendor_name()`.
+
+* Usage:
+
+  ```
+  show console model_name
+  ```
+
+* Example:
+
+  ```
+  admin@sonic:~$ show console model_name
+  Simulator
+  ```
+
+#### show console serial_number
+
+This command displays the serial number of the portable console device.
+
+* Usage:
+
+  ```
+  show console serial_number
```

Comment:
**From Xin's comments:**

Need to consider daisy-chain, we need to have all chained devices information.

Reply:
Thanks, I support daisy-chain in API design in this commit: [98aa6bc](https://github.com/sonic-net/SONiC/pull/1012/commits/98aa6bc59e7554fa18560b4ec341a626146521fc).